### PR TITLE
[CELEBORN-1075] Refactor `MetricsSystem` and `AbstractSource` to use synchronized blocks

### PR DESCRIPTION
**Pull request by cfmcgrady**
### What changes were proposed in this pull request?

As title

### Why are the changes needed?

Recently, during my testing on the main branch, we encountered `java.util.ConcurrentModificationException` error. This PR addresses synchronization issues in the MetricsSystem and AbstractSource classes by introducing synchronized blocks to ensure thread safety.

1. the `MetricsSystem#sources` collection has been changed from the `mutable.ArrayBuffer` to the `CopyOnWriteArrayList`, to prevent potential thread safety issues
2. the `AbstractSource#namedGauges` collection has been changed from the `ArrayList` to the `ConcurrentLinkedQueue` to enhance thread safety when adding gauges. to fix:

```
java.util.ConcurrentModificationException
        at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:911)
        at java.util.ArrayList$Itr.next(ArrayList.java:861)
        at scala.collection.convert.Wrappers$JIteratorWrapper.next(Wrappers.scala:46)
        at scala.collection.Iterator.foreach(Iterator.scala:943)
        at scala.collection.Iterator.foreach$(Iterator.scala:943)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
        at scala.collection.IterableLike.foreach(IterableLike.scala:74)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
        at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)
        at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)
        at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:184)
        at scala.collection.mutable.ListBuffer.$plus$plus$eq(ListBuffer.scala:47)
        at scala.collection.TraversableLike.to(TraversableLike.scala:786)
        at scala.collection.TraversableLike.to$(TraversableLike.scala:783)
        at scala.collection.AbstractTraversable.to(Traversable.scala:108)
        at scala.collection.TraversableOnce.toList(TraversableOnce.scala:350)
        at scala.collection.TraversableOnce.toList$(TraversableOnce.scala:350)
        at scala.collection.AbstractTraversable.toList(Traversable.scala:108)
        at org.apache.celeborn.common.metrics.source.AbstractSource.gauges(AbstractSource.scala:146)
        at org.apache.celeborn.common.metrics.source.AbstractSource.getMetrics(AbstractSource.scala:401)
        at org.apache.celeborn.common.metrics.sink.PrometheusServlet.$anonfun$getMetricsSnapshot$1(PrometheusServlet.scala:42)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
        at scala.collection.TraversableLike.map(TraversableLike.scala:286)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
        at scala.collection.AbstractTraversable.map(Traversable.scala:108)
        at org.apache.celeborn.common.metrics.sink.PrometheusServlet.getMetricsSnapshot(PrometheusServlet.scala:42)
        at org.apache.celeborn.common.metrics.sink.PrometheusHttpRequestHandler.handleRequest(PrometheusServlet.scala:59)
        at org.apache.celeborn.server.common.http.HttpRequestHandler.channelRead0(HttpRequestHandler.scala:53)
        at org.apache.celeborn.server.common.http.HttpRequestHandler.channelRead0(HttpRequestHandler.scala:37)
```

### Does this PR introduce _any_ user-facing change?

No, only bug fix

### How was this patch tested?

Pass GA

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala [hunk 1]
```diff
@@ -80,11 +80,14 @@ class MetricsSystem(
     MetricRegistry.name(source.sourceName)
   }
 
-  def getSourcesByName(sourceName: String): Seq[Source] =
+  def getSourcesByName(sourceName: String): Seq[Source] = sources.synchronized {
```
Comment (mridulm) [inline common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:83]: In `registerSinks`, this `sources` gets passed to `PrometheusServlet` - and so could be used outside of the `synchronized` lock.

Would have been more ideal if this was a thread safe variant instead (like `CopyOnWriteArrayList`).
Is it possible to update the constructor signature of `prometheusServlet` ?

Either to `List<Source>` (`sourcesList` directly) or atleast to `Seq[Source]` (`sourcesList.asScala`)

Comment (cfmcgrady) [reply common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:83]: SGTM, will update the constructor signature of `PrometheusServlet`.

Comment (cfmcgrady) [reply common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala]: addressed in https://github.com/apache/incubator-celeborn/pull/2023/commits/b5d7dff2e327757ca6bb386f950bf218f6692a32

File: common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala [hunk 2]
```diff
@@ -81,10 +81,10 @@ class MetricsSystem(
   }
 
   def getSourcesByName(sourceName: String): Seq[Source] =
-    sources.filter(_.sourceName == sourceName).toSeq
+    sources.asScala.filter(_.sourceName == sourceName).toSeq
```
Comment (mridulm) [inline common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:84]: nit: We can drop the `toSeq`

**Review by mridulm — APPROVED** [review]
Just a nit, looks good to me !

File: common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala (hunk 2 above)
Comment (cfmcgrady) [reply common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala]: updated.

Comment (cfmcgrady) [reply common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:84]: reverted, as the `toSeq` is required for Scala 2.13.

```
[error] /home/runner/work/incubator-celeborn/incubator-celeborn/common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:84:27: type mismatch;
[error]  found   : scala.collection.mutable.Buffer[org.apache.celeborn.common.metrics.source.Source]
[error]  required: Seq[org.apache.celeborn.common.metrics.source.Source]
[error]     sources.asScala.filter(_.sourceName == sourceName)
[error]                           ^
[error] one error found
```

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks! Merging to main(v0.4.0)/0.3(v0.3.2)

**Pull request closed without merging**